packages/core/test/BatchRenderer.js

```diff
@@ -1,43 +1,136 @@
-const { BatchRenderer } = require('../');
+const { Renderer, BatchRenderer, resources, Texture, BaseTexture } = require('../');
+const { CanvasResource } = resources;
+const { skipHello } = require('@pixi/utils');
 
-const mockrunner = {
-    contextChange: {
-        remove: () => 1,
-        add: () => 1,
-    },
-};
+skipHello();
 
-describe('BatchRenderer', function ()
+describe('PIXI.BatchRenderer', function ()
 {
-    it('can be destroyed', function ()
+    function createTexture(w, h)
     {
-        const destroyable = { destroy: sinon.stub() };
-        const webgl = {
-            on: sinon.stub(),
-            runners: mockrunner,
-            off: sinon.stub(),
-        };
-        const renderer = new BatchRenderer(webgl);
+        const canvas = document.createElement('canvas');
+        const context = canvas.getContext('2d');
 
-        // simulate onContextChange
-        renderer.vertexBuffers = [destroyable, destroyable];
-        renderer.vaos = [destroyable, destroyable];
-        renderer.indexBuffer = destroyable;
-        renderer.shader = destroyable;
+        canvas.width = w;
+        canvas.height = h;
+        context.fillStyle = 'white';
+        context.fillRect(0, 0, w, h);
 
-        expect(() => renderer.destroy()).to.not.throw();
-    });
+        return new BaseTexture(new CanvasResource(canvas));
+    }
 
-    it('can be destroyed immediately', function ()
+    it('should pass the batching test', function ()
```

Comment:
Wow, this looks huge, do you plan to have more batching tests in future?

Reply:
This test shown me a pair of stupid bugs, and it improves our coverage by more than 1%. I will make better architecture If I ever need second test.

Reply:
With better architecture I already know a number of tests I want to add. However, I've spent too many hours on this PR to go further :)

Reply:
Ok, let's add more in the future.